File: Mavim-services/Topic/Mavim.Manager.Api.Topic.Repository/v1/Mappers/NumberFieldMapper.cs
using Mavim.Manager.Api.Topic.Repository.Interfaces.v1.enums;
using Mavim.Manager.Api.Topic.Repository.Interfaces.v1.Fields;
using Mavim.Manager.Api.Topic.Repository.v1.Fields;
using Mavim.Manager.Api.Topic.Repository.v1.Mappers.Abstract;
using System;

namespace Mavim.Manager.Api.Topic.Repository.v1.Mappers
{
    internal class NumberFieldMapper : GenericNumberFieldMapper<Model.ISingleValueSimpleField, ISingleNumberField>
    {
        /// <summary>
        /// Gets the generic mapped number service field.
        /// </summary>
        /// <param name="field">The field.</param>
        /// <returns></returns>
        /// <exception cref="ArgumentNullException">field</exception>
        protected override ISingleNumberField GetGenericMappedServiceField(Model.ISingleValueSimpleField field)
        {
            if (field == null) throw new ArgumentNullException(nameof(field));

            return new SingleNumberField
            {
                FieldId = field.FieldDefinition.ID,
                FieldName = field.FieldDefinition.Name,
                SetOrder = field.FieldSetDefinition.OrderNumber,
                Order = field.OrderNumber,
                FieldSetId = field.FieldSetDefinition.ID,
                SetName = field.FieldSetDefinition?.Name ?? string.Empty,
                FieldValue = field.FieldValue?.ToString(),
                FieldValueType = FieldType.Number,
                Required = field.FieldDefinition.Required,
                Readonly = field.ReadOnly
            };
        }

        /// <summary>
        /// Gets the generic mapped number repo field.
        /// </summary>
        /// <param name="field">The field.</param>
        /// <param name="simpleField">The simple field.</param>
        /// <returns></returns>
        /// <exception cref="ArgumentNullException">field</exception>
        protected override object[] GetGenericMappedRepoField(ISingleNumberField field, Model.ISimpleField simpleField = null)
        {
            return new object[] { field?.FieldValue };
        }
    }
}

File: Mavim-apps/Mavim.frontend/projects/mavim-app/src/shared/field/mappers/numberfield.mapper.ts
import { Field } from '../models/field.model';
import { SingleNumberField } from '../models/api';
import { GenericNumberFieldMapper } from './abstract/generic-number-field.mapper';
import { mapToApiFieldType, mapToFieldType } from './fieldvaluetype';

export class NumberFieldMapper extends GenericNumberFieldMapper<SingleNumberField> {
	public get getUrlTypeName(): string {
		return 'number';
	}

	public getMappedField(field: SingleNumberField): Field {
		return this.mapField(field);
	}

	public getMappedApiField(field: Field): SingleNumberField {
		return this.mapApiField(field);
	}

	protected mapApiField(field: Field): SingleNumberField {
		return {
			...field,
			fieldValueType: mapToApiFieldType(field.fieldValueType),
			data: field.data ? this.mapToNumber(field.data[0]) : undefined
		} as SingleNumberField;
	}

	protected mapField(field: SingleNumberField): Field {
		return {
			...field,
			fieldValueType: mapToFieldType(field.fieldValueType),
			options: undefined,
			data: [this.mapToString(field.data)]
		} as Field;
	}
}
